### PR TITLE
added EPD3in7

### PR DESCRIPTION
**Pull request by sonic74**

### CODE REVIEW

**Comment by joukos** [comment]
Thanks for the contribution! I don't have this display so can't really test it myself, though.

So just to be clear, this driver provides B/W functionality but grayscale would be supported by the display if it was implemented?

**Comment by sonic74** [comment]
Indeed! But it would get even slower.

File: papertty/drivers/drivers_full.py [hunk 1]
```diff
@@ -270,6 +270,177 @@ def sleep(self):
...
+        0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 
+        0x00,0x00,0x03,0x05,0x00,0x00,0x00,0x00,0x00,0x00, 
+        0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 
+        0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 
+        0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 
+        0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 
+        0x22,0x22,0x22,0x22,0x22
+    ]
+
+    def ReadBusy(self):
```
Comment (joukos) [inline papertty/drivers/drivers_full.py:323]: For consistency, should rename this method (note to self - maybe refactor base `wait_until_idle` to take parameters to cater for all variants so it doesn't need to be overridden):

```suggestion
    def wait_until_idle(self):
```

File: papertty/drivers/drivers_full.py [hunk 2]
```diff
@@ -270,6 +270,177 @@ def sleep(self):
...
+            self.send_command(0x4F)
+            self.send_data(0x00)
+            self.send_data(0x00)
+
+            self.send_command(0x24)
+            for j in range(0, self.height):
+                for i in range(0, int(self.width / 8)):
+                    self.send_data(frame_buffer[i + j * int(self.width / 8)])   
+
+#            self.load_lut(self.lut_1Gray_A2)
```
Comment (joukos) [inline papertty/drivers/drivers_full.py:424]: I guess this comment can be removed.

```suggestion
```

File: papertty/drivers/drivers_full.py [hunk 3]
```diff
@@ -270,6 +270,177 @@ def sleep(self):
         self.send_data(0xa5)
 
 
+class EPD3in7(WaveshareFull):
+    """Waveshare 3.7" - monochrome"""
+
+    def __init__(self):
+        super().__init__(name='3.7" BW', width=280, height=480)
+
+    # not ported: 4-greyscale mode, partial refresh (~600-900 ms, EPD_3in7.c)
```
Comment (joukos) [inline papertty/drivers/drivers_full.py:279]: It's unclear which file is referenced - assuming related to Waveshare sample code. Could be bit more verbose here (and reference the Python source instead):
```suggestion
    # The 3.7" display hardware supports B/W and 4-shade grayscale modes, but the
    # grayscale mode is not yet implemented here. Reference code in Waveshare's repo:
    # https://github.com/waveshare/e-Paper/blob/master/RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd3in7.py
    # Updating the screen in grayscale mode takes approximately 600-900 ms.
```

**Review by joukos — CHANGES_REQUESTED** [review]
A few minor suggestions.

File: papertty/drivers/drivers_full.py (hunk 2 above)
Comment (sonic74) [reply papertty/drivers/drivers_full.py:424]: It was supposed to document the original call order.

**Review by joukos — APPROVED** [review]

**Comment by joukos** [comment]
Thanks, merged this now - I'll try to update the PyPi package when I have a bit of time.

**Pull request merged**